client/x/evmstaking/keeper/abci.go

```diff
@@ -28,99 +28,125 @@ func (k *Keeper) EndBlock(ctx context.Context) (abci.ValidatorUpdates, error) {
 		if err != nil {
-			return nil, errors.Wrap(err, "validator address from bech32")
+			return nil, err
 		}
 
-		delegatorAddr, err := k.authKeeper.AddressCodec().StringToBytes(dvPair.DelegatorAddress)
-		if err != nil {
-			return nil, errors.Wrap(err, "delegator address from bech32")
-		}
+		// make an array with each entry being the validator address, delegator address, and the amount
+		var unbondedEntries []UnbondedEntry
 
-		ubd, err := (k.stakingKeeper).GetUnbondingDelegation(ctx, delegatorAddr, validatorAddr)
-		if err != nil {
-			return nil, err
-		}
+		for _, dvPair := range matureUnbonds {
+			validatorAddr, err := k.validatorAddressCodec.StringToBytes(dvPair.ValidatorAddress)
+			if err != nil {
+				return nil, errors.Wrap(err, "validator address from bech32")
+			}
 
-		// TODO: parameterized bondDenom
-		bondDenom := sdk.DefaultBondDenom
-
-		// loop through all the entries and process unbonding mature entries
-		for i := range len(ubd.Entries) {
-			entry := ubd.Entries[i]
-			if entry.IsMature(ctxTime) && !entry.OnHold() {
-				// track undelegation only when remaining or truncated shares are non-zero
-				if !entry.Balance.IsZero() {
-					amt := sdk.NewCoin(bondDenom, entry.Balance)
-					// TODO: check if it's possible to add a double entry in the unbondedEntries array
-					unbondedEntries = append(unbondedEntries, UnbondedEntry{
-						validatorAddress: dvPair.ValidatorAddress,
-						delegatorAddress: dvPair.DelegatorAddress,
-						amount:           amt.Amount,
-					})
-				}
+			delegatorAddr, err := k.authKeeper.AddressCodec().StringToBytes(dvPair.DelegatorAddress)
+			if err != nil {
+				return nil, errors.Wrap(err, "delegator address from bech32")
 			}
-		}
-	}
 
-	valUpdates, err := k.stakingKeeper.EndBlocker(ctx)
-	if err != nil {
-		return nil, err
-	}
+			ubd, err := (k.stakingKeeper).GetUnbondingDelegation(ctx, delegatorAddr, validatorAddr)
+			if err != nil {
+				return nil, err
+			}
 
-	for _, entry := range unbondedEntries {
-		log.Debug(ctx, "Adding undelegation to withdrawal queue",
-			"delegator", entry.delegatorAddress,
-			"validator", entry.validatorAddress,
-			"amount", entry.amount.String())
+			// TODO: parameterized bondDenom
+			bondDenom := sdk.DefaultBondDenom
+
+			// loop through all the entries and process unbonding mature entries
+			for i := range len(ubd.Entries) {
+				entry := ubd.Entries[i]
+				if entry.IsMature(ctxTime) && !entry.OnHold() {
+					// track undelegation only when remaining or truncated shares are non-zero
+					if !entry.Balance.IsZero() {
+						amt := sdk.NewCoin(bondDenom, entry.Balance)
+						// TODO: check if it's possible to add a double entry in the unbondedEntries array
+						unbondedEntries = append(unbondedEntries, UnbondedEntry{
+							validatorAddress: dvPair.ValidatorAddress,
+							delegatorAddress: dvPair.DelegatorAddress,
+							amount:           amt.Amount,
+						})
```

Comment:
I thought we have a limit on the number of operations per unbonding period or epoch. Do we know where that is enforced?

Reply:
As commented [here](https://github.com/piplabs/story/pull/96#discussion_r1794884092), no limit in this PR. Will do in the following PR.